src/engine/strat_engine/range_alloc.rs

```diff
@@ -43,10 +41,52 @@ impl RangeAllocator {
         for &(off, len) in ranges {
             self.check_for_overflow(off, len);
 
-            for val in *off..*off + *len {
-                let inserted = self.used.insert(val);
-                if inserted == false {
-                    panic!(format!("inserted value {} already present", val));
+            let contig_prev = self.used
+                .range(..off)
+                .rev()
+                .next()
+                .map(|(k, v)| (*k, *v))
+                .and_then(|(prev_off, prev_len)| {
+                    assert!(off >= prev_off + prev_len,
+                            "given range must not overlap previous range");
+                    if prev_off + prev_len == off {
+                        Some((prev_off, prev_len))
+                    } else {
+                        None
+                    }
+                });
+
+            let contig_next = self.used
+                .range(off..)
+                .next()
+                .map(|(k, v)| (*k, *v))
+                .and_then(|(next_off, next_len)| {
+                    assert!(off + len >= next_off,
```

Comment:
I'm finding it hard to convince myself that this shouldn't be ```<=``` (not ```>=```). If I'm right, it's probably worth another test.